src/charm.py

```diff
@@ -252,12 +259,29 @@ def __init__(self, *args):
             self._loki_push_api_alert_rules_changed,
         )
         self.framework.observe(self.on.logging_relation_changed, self._on_logging_relation_changed)
+
+        self.framework.observe(
+            self.on.send_datasource_relation_changed, self._on_grafana_source_changed
```

Comment:
off the top of my head, we should only be updating DSX on grafana-source-relation-changed/departed (perhaps broken too?)

if dsx itself changes or departs, we really don't care as loki won't need to do anything about it.

Reply:
currently, yes, loki won't need to do anything
I put it there for when loki would actually be doing some correlation itself
will remove it for now

Reply:
no wait
if we only have grafana-source relation and no dsx and we later add that dsx relation, loki will never publish. That's why we need to observe on dsx as well